op-batcher/batcher/driver.go

```diff
@@ -901,8 +901,19 @@ func (l *BatchSubmitter) recordFailedTx(id txID, err error) {
 
 func (l *BatchSubmitter) recordConfirmedTx(id txID, receipt *types.Receipt) {
 	l.Log.Info("Transaction confirmed", logFields(id, receipt)...)
-	l1block := eth.ReceiptBlockID(receipt)
-	l.state.TxConfirmed(id, l1block)
+	tctx, cancel := context.WithTimeout(l.shutdownCtx, l.Config.NetworkTimeout)
+	defer cancel()
+
+	var header *types.Header
+	var err error
+	for {
+		header, err = l.L1Client.HeaderByNumber(tctx, receipt.BlockNumber)
+		if err != nil {
+			break
+		}
```

Comment:
Don't you need to break on `nil` error? We should also warn-log that we reattempt and put a pause or this can be a busy loop.
```suggestion
		if err == nil {
			break
		}
		l.log.Warn("Getting header for receipt failed, retrying.", "err", err)
		time.Sleep(2 * time.Second)
```
Wondering if there's also a generic function for this in the `retry` package.

I'm also wondering if there's an edge case where we receive a receipt, then an L1 reorg occurs and then we're getting the block time for a different block (but same block number), that might not include the receipt. We could fetch by block hash instead, but then the reorg case would fail here indefinitely...

Maybe the best we can do here is get by number, as it is now, but then check the block hash and in case it's different, we'd not confirm the tx but error-log and return early. However, the new block may also contain another receipt for the same transaction. We should check whether we'd get that new receipt as well by the txmgr logic. Then this code would be called again for the new receipt and we can safely just return in such a reorg case.

Reply:
Good catch, fixed here [6d9adff](https://github.com/ethereum-optimism/optimism/pull/12743/commits/6d9adffd707be5b02eb838509e9ec906ddda85c0). 

The retry package does not allow us to retry indefinitely. We can either roll our own logic here or perhaps modify the retry package to have an `Indefinite` strategy.

From what I can tell about the txmgr: it will only return one receipt per transaction. So if that tx gets included, a receipt will be sent and then if there is an L1 reorg and the tx is included again (without being sent again by the txmgr) then there would not be a second receipt event in the batcher.